### PR TITLE
test: Setup jenkins ci

### PR DESCRIPTION
**Pull request by mxinden**
Automate prometheus-operator e2e tests via Jenkins.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -5,7 +5,7 @@ NAMESPACE?=prometheus-operator-e2e-tests-$(shell LC_CTYPE=C tr -dc a-z0-9 < /dev
 PROMU := $(GOPATH)/bin/promu
 PREFIX ?= $(shell pwd)
 
-CLUSTER_IP?=$(shell minikube ip)
+CLUSTER_IP?=$(shell kubectl config view --minify | grep server: | cut -f 3 -d ":" | tr -d "//")
```
Comment (brancz) [inline Makefile]: how can we make sure that we locally don't accidentally run the tests against a production cluster as that happens to be the context that is currently selected?

File: scripts/jenkins/ed.sh [hunk 2]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
```
Comment (brancz) [inline scripts/jenkins/ed.sh]: maybe explain why this script is required, so we in the future or someone new can understand this hack

File: scripts/jenkins/wait-for-cluster.sh [hunk 3]
```diff
@@ -0,0 +1,4 @@
+until kubectl cluster-info
```
Comment (brancz) [inline scripts/jenkins/wait-for-cluster.sh]: missing she-bang and it's always good to have a quick description of what this script does ... generally following https://google.github.io/styleguide/shell.xml

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                credentialsId('Jenkins-Monitoring-AWS-User')
+            }
+            usernamePassword('QUAY_ROBOT_USERNAME', 'QUAY_ROBOT_SECRET', 'quay_robot')
+        }
+    }
+
+    triggers {
+        githubPullRequest {
+           useGitHubHooks()
+            admins(['mxinden'])
```
Comment (brancz) [inline Jenkinsfile]: this looks like it's a leftover from testing, is this still required?

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            orgWhitelist(['coreos-inc'])
+        }
+    }
+
+    steps {
+        shell('docker build -t cluster-setup-env scripts/jenkins/.')
+    }
+
+    steps {
+        shell('docker run --privileged --rm -v $PWD:/go/src/github.com/coreos/prometheus-operator/ cluster-setup-env /bin/bash -c "cd /go/src/github.com/coreos/prometheus-operator && CGO_ENABLED=0 GOOS=linux go build --ldflags \'-extldflags \\"-static\\"\' github.com/coreos/prometheus-operator/cmd/operator"')
```
Comment (brancz) [inline Jenkinsfile]: we already discussed this offline and makes sense to only address this in a follow up PR, but we should work towards replacing this with `promu`

File: scripts/jenkins/clusters/prom-test-indenml/spec.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+awsRegion: eu-west-1
+prometheus:
+  machineType: d2.medium
+  scrapeInterval: 4s
+  instances:
```
Comment (brancz) [inline scripts/jenkins/clusters/prom-test-indenml/spec.yaml]: Is this intended to be blank?

File: scripts/jenkins/clusters/prom-test-indenml/spec.yaml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+awsRegion: eu-west-1
```
Comment (brancz) [inline scripts/jenkins/clusters/prom-test-indenml/spec.yaml]: Is this yaml still required? If yes the folder should probably be renamed.

File: scripts/jenkins/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,29 @@
+FROM golang:1.7.5-wheezy
+
+ENV TERRAFORM_VERSION 0.8.4
```
Comment (brancz) [inline scripts/jenkins/Dockerfile]: Hoping this does not make a difference but `0.8.7` is latest.

File: scripts/jenkins/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+RUN curl "https://s3.amazonaws.com/aws-cli/awscli-bundle.zip" -o "awscli-bundle.zip" && \
+    unzip awscli-bundle.zip && \
+    ./awscli-bundle/install -i /usr/local/aws -b /bin/aws
+
+RUN curl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl && \
+    chmod +x ./kubectl && \
+    mv ./kubectl /bin/kubectl
+
+RUN curl -L https://get.docker.com/builds/Linux/x86_64/docker-1.6.0 > /usr/local/bin/docker && \
```
Comment (brancz) [inline scripts/jenkins/Dockerfile]: Why this version? Also should be moved to the env vars above.

File: scripts/jenkins/wait-for-cluster.sh (hunk 3 above)
Comment (brancz) [inline scripts/jenkins/wait-for-cluster.sh]: You'll also want a `set -e` here to make the script exit with a non 0 exit code if say `kubectl cluster-info` errors.

File: scripts/jenkins/ed.sh (hunk 2 above)
Comment (brancz) [inline scripts/jenkins/ed.sh]: Additionally we probably want to validate the input at least a bit, to protect against the Makefile accidentally writing somewhere crazy here and to make sure the source actually exists.

Comment (brancz) [inline scripts/jenkins/ed.sh]: And if it's more than just `cat $1 > $2` we probably want to name the variables to express their content.

File: Makefile (hunk 1 above)
Comment (mxinden) [inline Makefile]: Resolved for now

Comment (brancz) [inline Makefile]: For the reference: may want to run locally against other clusters than minikube therefore a simple "if CI" won't work, so skipping this for now.

**Comment by mxinden** [comment]
test this please

File: scripts/jenkins/ed.sh [hunk 10]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+# Kops requires user input through an editor. This shell script simply returns
+# the file input '$1' to STDOUT which enables us to script editing nodes through
```
Comment (brancz) [inline scripts/jenkins/ed.sh]: I'm not sure this is 100% correct. Maybe @alexsomesan can comment. IIRC how this hack works is that kops calls `$EDITOR current-state-of-file after-edit-done-file`. And as we can pass the first one directly to kops and we are sure this is what we want as it has been checked in to git we can just write it to the "destination" file.

Either way naming $1 and $2 would definitely help.

Comment (brancz) [inline scripts/jenkins/ed.sh]: Looking at how this is used it's actually that we produce a fake editor for every "edit" with kops that always just writes the same file (`$1`) to the "to-edit" file (`$2`), which is probably some tmp file kops controls.

**Comment by mxinden** [comment]
test this please

File: scripts/jenkins/wait-for-cluster.sh (hunk 3 above)
Comment (mxinden) [inline scripts/jenkins/wait-for-cluster.sh]: resolved offline.

**Comment by mxinden** [comment]
test this please

File: scripts/jenkins/wait-for-cluster.sh (hunk 3 above)
Comment (brancz) [inline scripts/jenkins/wait-for-cluster.sh]: Always good to documented anyways what the result of the offline discussion was for everyone else to understand and for us to understand when we go back.

For the reference: there was only the `until` statement and the `sleep` function so there was nothing to error.

However, as this script is now more than that, I propose that `set -e` is added. Plus, even if we don't see a possible error condition doesn't mean there isn't one :wink:

File: Jenkinsfile [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        shell('docker build -t cluster-setup-env scripts/jenkins/.')
+    }
+
+    steps {
+        shell('docker run --rm -v /var/jenkins/workspace/e2e-playground:/var/jenkins/workspace/e2e-playground -v /var/run/docker.sock:/var/run/docker.sock cluster-setup-env /bin/bash -c "cd /var/jenkins/workspace/e2e-playground && make crossbuild"')
+    }
+
+    steps {
+        shell('docker build -t quay.io/coreos/prometheus-operator-dev:$BUILD_ID .')
+        shell('docker login -u="$QUAY_ROBOT_USERNAME" -p="$QUAY_ROBOT_SECRET" quay.io')
```
Comment (glerchundi) [inline Jenkinsfile]: as this is going to be expanded, the credentials will be leaked in the build log, right?

Comment (fabxc) [inline Jenkinsfile]: Very good point, thanks for bringing that up! @mxinden can you check this?

Comment (mxinden) [inline Jenkinsfile]: @glerchundi @fabxc As we are using the Jenkins credentials plugin it automatically masks them in the log output: `docker login -u=**** -p=**** quay.io`. Thank you for paying attention though! Good feedback.

**Comment by mxinden** [comment]
test this please

**Comment by mxinden** [comment]
@brancz Could you re-review this PR?

File: Jenkinsfile [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+                secretKeyVariable('AWS_SECRET_ACCESS_KEY')
+                credentialsId('Jenkins-Monitoring-AWS-User')
+            }
+            usernamePassword('QUAY_ROBOT_USERNAME', 'QUAY_ROBOT_SECRET', 'quay_robot')
+        }
+    }
+
+    triggers {
+        githubPullRequest {
+           useGitHubHooks()
```
Comment (brancz) [inline Jenkinsfile]: everything else is indented with 4 spaces, any reason this is not?

File: scripts/jenkins/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+ENV TERRAFORM_VERSION 0.8.7
+ENV KOPS_VERSION 1.5.1
+ENV DOCKER_VERSION 1.13.1
+
+RUN echo "deb http://ftp.debian.org/debian wheezy-backports main" >> /etc/apt/sources.list
+
+RUN apt-get update && apt-get install -y --no-install-recommends \
+		ca-certificates \
+		wget \
```
Comment (brancz) [inline scripts/jenkins/Dockerfile]: tabs -> spaces

File: Jenkinsfile (hunk 11 above)
Comment (brancz) [inline Jenkinsfile]: phew, quite a lot of trust we're putting in that plugin, but I guess this is ok

File: Jenkinsfile [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+                credentialsId('Jenkins-Monitoring-AWS-User')
+            }
+            usernamePassword('QUAY_ROBOT_USERNAME', 'QUAY_ROBOT_SECRET', 'quay_robot')
+        }
+    }
+
+    triggers {
+        githubPullRequest {
+            useGitHubHooks()
+                orgWhitelist(['coreos-inc'])
```
Comment (brancz) [inline Jenkinsfile]: is this indentation correct? it looks like two function calls one after another to me, so they should be aligned, no?

**Comment by brancz** [comment]
lgtm :+1: nice job!

File: Jenkinsfile (hunk 11 above)
Comment (glerchundi) [inline Jenkinsfile]: thanks to you guys for such a good project 👍

**Pull request merged**